## Introdução a análise de dados com Pandas - Parte 4
É possível nos depararmos com dados incompletos ou com erros de registro. O pacote pandas pode ser utilizado para tratar os dados antes de iniciar o processo de análise visual dos dados.

In [1]:
import pandas as pd
import matplotlib as plt

Para experimentar alguns comandos, utilizaremos a base "housing.csv". Ao contrário da base "housing_ok.csv", a primeira apresenta dados em branco e outros com valores incoerentes com a realidade.

In [2]:
housing = pd.read_csv('housing.csv')

In [3]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11306 entries, 0 to 11305
Data columns (total 14 columns):
Bairro                    11304 non-null object
Endereco                  11304 non-null object
Quartos                   11304 non-null float64
Tipo                      11304 non-null object
Preco                     8827 non-null float64
Distancia do aeroporto    11304 non-null float64
CEP                       11304 non-null float64
Banheiro                  11015 non-null float64
Vagas                     11055 non-null float64
Area terreno              9031 non-null float64
Area construida           6104 non-null float64
Ano da construcao         7016 non-null float64
Latitude                  11306 non-null object
Longitude                 11092 non-null object
dtypes: float64(9), object(5)
memory usage: 1.2+ MB


Através do comando "info()" é possível observar alguns problemas.
1. Quase todas as colunas possuem dados em branco;
2. As colunas Latitude e Longitude foram interpretadas como tipo "object"

Trataremos primeiro os dados em branco.
Podemos (basicamente) lidar com dados em brancos de duas formas:
* Eliminar linhas que contanham um ou mais dados em branco
* Substituir os dados em branco por valores quaisquer

Geralmente a primeira opção é a mais adequada.
Para eliminar todas as linhas contendo pelo menos um valor em branco o utilizamos o comando "dropnan".

In [4]:
housing.dropna(axis=0, inplace=True)

O parâmetro "axis=0" indica que faremos a eliminação de linhas e não colunas (axis=1 para coluna)

Observando a tabela após a eliminação de linhas:

In [5]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3627 entries, 1 to 7015
Data columns (total 14 columns):
Bairro                    3627 non-null object
Endereco                  3627 non-null object
Quartos                   3627 non-null float64
Tipo                      3627 non-null object
Preco                     3627 non-null float64
Distancia do aeroporto    3627 non-null float64
CEP                       3627 non-null float64
Banheiro                  3627 non-null float64
Vagas                     3627 non-null float64
Area terreno              3627 non-null float64
Area construida           3627 non-null float64
Ano da construcao         3627 non-null float64
Latitude                  3627 non-null object
Longitude                 3627 non-null object
dtypes: float64(9), object(5)
memory usage: 425.0+ KB


Passamos de 11306 linhas para 3627. Redução de quase 70% da base.

Agora faremos a transformação das colunas "Latitude" e "Longitude" de "object" (texto) para "float64" (número) utilizando o comando "pd.to_numeric()".

In [6]:
housing.Longitude = pd.to_numeric(housing.Longitude, errors='coerce')
housing.Latitude = pd.to_numeric(housing.Latitude, errors='coerce')

O comando "pd.to_numeric()" recebe dois parâmetros: coluna e a forma de lidar com erros. 
A configuração erros='coerce' significa que o conversor irá forçar a conversão.
Observando as informações da tabela, é possível observar que a conversão foi feita com sucesso.

In [7]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3627 entries, 1 to 7015
Data columns (total 14 columns):
Bairro                    3627 non-null object
Endereco                  3627 non-null object
Quartos                   3627 non-null float64
Tipo                      3627 non-null object
Preco                     3627 non-null float64
Distancia do aeroporto    3627 non-null float64
CEP                       3627 non-null float64
Banheiro                  3627 non-null float64
Vagas                     3627 non-null float64
Area terreno              3627 non-null float64
Area construida           3627 non-null float64
Ano da construcao         3627 non-null float64
Latitude                  3627 non-null float64
Longitude                 3626 non-null float64
dtypes: float64(11), object(3)
memory usage: 425.0+ KB


Para investigar possíveis anomalias nos valores das colunas, utilizaremos o comando "describe()"

In [8]:
housing.describe()

Quartos         Preco  Distancia do aeroporto          CEP  \
count  3627.000000  3.627000e+03             3627.000000  3627.000000   
mean      3.092914  1.215652e+06                8.972594  3091.642404   
std       0.901321  7.017057e+05                3.466525    58.874382   
min       1.000000  1.310000e+05                1.200000  3000.000000   
25%       3.000000  7.411250e+05                6.400000  3042.000000   
50%       3.000000  1.025000e+06                9.200000  3073.000000   
75%       4.000000  1.500000e+06               11.700000  3143.000000   
max       8.000000  8.000000e+06               15.000000  3207.000000   

          Banheiro        Vagas  Area terreno  Area construida  \
count  3627.000000  3627.000000   3627.000000      3627.000000   
mean      1.631100     1.609319    520.160463       153.382534   
std       0.743868     0.955635    653.564225        95.890805   
min       1.000000     0.000000     43.000000         1.000000   
25%       1.000000     1.000000    227.000000       102.000000   
50%       2.000000     2.000000    449.000000       132.000000   
75%       2.000000     2.000000    648.000000       180.000000   
max       8.000000     8.000000  15900.000000      3112.000000   

       Ano da construcao      Latitude      Longitude  
count        3627.000000   3627.000000    3626.000000  
mean         1956.990074  -3254.972992   13407.429455  
std            38.744626  10543.701635   41780.754937  
min          1830.000000 -37946.000000     144.791800  
25%          1925.000000    -37.867000     144.931700  
50%          1960.000000    -37.805500     145.002200  
75%          1995.000000    -37.762700     145.068800  
max          2018.000000    -37.686200  145138.000000

Observa-se uma divergência nos valores da Latitude e Longitude.
No valor mínimo da Latitude aparece um valor muito menor do que a média, indicando erro de cadastro. O mesmo acontece com o valor máximo da Longitude.

O erro de cadastro, claramente, envolveu a ausência do divisor decimal. Para corrigir, faremos a divisão dos valores da Latitude e Longitude que estão com esse problema por 1000.

In [9]:
housing.loc[housing.Latitude < -1000, 'Latitude'] = housing.Latitude / 1000
housing.loc[housing.Longitude > 1000, 'Longitude'] = housing.Longitude / 1000

Utilizando o comando "info()" observamos que a divergência foi corrigida.

In [10]:
housing.describe()

Quartos         Preco  Distancia do aeroporto          CEP  \
count  3627.000000  3.627000e+03             3627.000000  3627.000000   
mean      3.092914  1.215652e+06                8.972594  3091.642404   
std       0.901321  7.017057e+05                3.466525    58.874382   
min       1.000000  1.310000e+05                1.200000  3000.000000   
25%       3.000000  7.411250e+05                6.400000  3042.000000   
50%       3.000000  1.025000e+06                9.200000  3073.000000   
75%       4.000000  1.500000e+06               11.700000  3143.000000   
max       8.000000  8.000000e+06               15.000000  3207.000000   

          Banheiro        Vagas  Area terreno  Area construida  \
count  3627.000000  3627.000000   3627.000000      3627.000000   
mean      1.631100     1.609319    520.160463       153.382534   
std       0.743868     0.955635    653.564225        95.890805   
min       1.000000     0.000000     43.000000         1.000000   
25%       1.000000     1.000000    227.000000       102.000000   
50%       2.000000     2.000000    449.000000       132.000000   
75%       2.000000     2.000000    648.000000       180.000000   
max       8.000000     8.000000  15900.000000      3112.000000   

       Ano da construcao     Latitude    Longitude  
count        3627.000000  3627.000000  3626.000000  
mean         1956.990074   -37.805456   144.986199  
std            38.744626     0.062661     0.078533  
min          1830.000000   -37.946000   144.791800  
25%          1925.000000   -37.849050   144.925125  
50%          1960.000000   -37.796600   144.995300  
75%          1995.000000   -37.758635   145.050300  
max          2018.000000   -37.686200   145.143600